icicle/utils/vec_ops.cu

```diff
@@ -31,165 +33,115 @@ namespace vec_ops {
-      CHK_IF_RETURN(cudaFreeAsync(d_vec_b, ctx.stream));
-      CHK_IF_RETURN(cudaFreeAsync(d_result, ctx.stream));
+    if (!config.is_b_on_device) {
+      CHK_IF_RETURN(cudaMallocAsync(&d_vec_b, n * sizeof(E), config.ctx.stream));
+      CHK_IF_RETURN(cudaMemcpyAsync(d_vec_b, vec_b, n * sizeof(E), cudaMemcpyHostToDevice, config.ctx.stream));
+    } else {
+      d_vec_b = vec_b;
     }
 
-    return CHK_STICKY(cudaStreamSynchronize(ctx.stream));
-  }
-
-  template <typename E>
-  cudaError_t Add(E* vec_a, E* vec_b, int n, bool is_on_device, device_context::DeviceContext ctx, E* result)
-  {
-    CHK_INIT_IF_RETURN();
+    if (!config.is_result_on_device) {
+      CHK_IF_RETURN(cudaMallocAsync(&d_result, n * sizeof(E), config.ctx.stream));
+    } else {
+      d_result = result;
+    }
 
-    // Set the grid and block dimensions
-    int num_threads = MAX_THREADS_PER_BLOCK;
-    int num_blocks = (n + num_threads - 1) / num_threads;
+    // Call the kernel to perform element-wise operation
+    Kernel<<<num_blocks, num_threads, 0, config.ctx.stream>>>(d_vec_a, d_vec_b, n, d_result);
+    if (config.is_result_montgomery_form) CHK_IF_RETURN(mont::FromMontgomery(d_result, n, config.ctx.stream, d_result));
 
-    E *d_vec_a, *d_vec_b, *d_result;
-    if (!is_on_device) {
-      // Allocate memory on the device for the input vectors and the output vector
-      CHK_IF_RETURN(cudaMallocAsync(&d_vec_a, n * sizeof(E), ctx.stream));
-      CHK_IF_RETURN(cudaMallocAsync(&d_vec_b, n * sizeof(E), ctx.stream));
-      CHK_IF_RETURN(cudaMallocAsync(&d_result, n * sizeof(E), ctx.stream));
-
-      // Copy the input vectors from the host to the device
-      CHK_IF_RETURN(cudaMemcpyAsync(d_vec_a, vec_a, n * sizeof(E), cudaMemcpyHostToDevice, ctx.stream));
-      CHK_IF_RETURN(cudaMemcpyAsync(d_vec_b, vec_b, n * sizeof(E), cudaMemcpyHostToDevice, ctx.stream));
-    }
+    if (!config.is_a_on_device) { CHK_IF_RETURN(cudaFreeAsync(d_vec_a, config.ctx.stream)); }
 
-    // Call the kernel to perform element-wise addition
-    AddKernel<<<num_blocks, num_threads, 0, ctx.stream>>>(
-      is_on_device ? vec_a : d_vec_a, is_on_device ? vec_b : d_vec_b, n, is_on_device ? result : d_result);
+    if (!config.is_b_on_device) { CHK_IF_RETURN(cudaFreeAsync(d_vec_b, config.ctx.stream)); }
 
-    if (!is_on_device) {
-      CHK_IF_RETURN(cudaMemcpyAsync(result, d_result, n * sizeof(E), cudaMemcpyDeviceToHost, ctx.stream));
-      CHK_IF_RETURN(cudaFreeAsync(d_vec_a, ctx.stream));
-      CHK_IF_RETURN(cudaFreeAsync(d_vec_b, ctx.stream));
-      CHK_IF_RETURN(cudaFreeAsync(d_result, ctx.stream));
+    if (!config.is_result_on_device) {
+      CHK_IF_RETURN(cudaMemcpyAsync(result, d_result, n * sizeof(E), cudaMemcpyDeviceToHost, config.ctx.stream));
+      CHK_IF_RETURN(cudaFreeAsync(d_result, config.ctx.stream));
     }
 
-    return CHK_STICKY(cudaStreamSynchronize(ctx.stream));
+    return CHK_STICKY(cudaStreamSynchronize(config.ctx.stream));
```

Comment:
Shouldn't syncing be dependent on `config.is_async`?

Reply:
of course. Good catch